Firestore/core/src/api/load_bundle_task.h

```diff
@@ -0,0 +1,197 @@
+                         uint32_t total_documents,
+                         uint64_t bytes_loaded,
+                         uint64_t total_bytes,
+                         LoadBundleTaskState state)
+      : documents_loaded_(documents_loaded),
+        total_documents_(total_documents),
+        bytes_loaded_(bytes_loaded),
+        total_bytes_(total_bytes),
+        state_(state) {
+  }
+
+  /** Returns how many documents have been loaded. */
+  uint32_t documents_loaded() const {
+    return documents_loaded_;
+  }
+
+  /**
+   * Returns the total number of documents in the bundle. Returns 0 if the
+   * bundle failed to parse.
+   */
+  uint32_t total_documents() const {
+    return total_documents_;
+  }
+
+  /** Returns how many bytes have been loaded. */
+  uint64_t bytes_loaded() const {
+    return bytes_loaded_;
+  }
+
+  /** Returns the total number of bytes in the bundle. Returns 0 if the bundle
+   * failed to parse. */
+  uint64_t total_bytes() const {
+    return total_bytes_;
+  }
+
+  /** Returns the current state of the task. */
+  LoadBundleTaskState state() const {
+    return state_;
+  }
+
+  void set_state(LoadBundleTaskState state) {
+    state_ = state;
+  }
+
+ private:
+  uint32_t documents_loaded_ = 0;
+  uint32_t total_documents_ = 0;
+  uint64_t bytes_loaded_ = 0;
+  uint64_t total_bytes_ = 0;
+
+  LoadBundleTaskState state_ = LoadBundleTaskState::InProgress;
+};
+
+inline bool operator==(const LoadBundleTaskProgress lhs,
+                       const LoadBundleTaskProgress& rhs) {
+  return lhs.state() == rhs.state() &&
+         lhs.bytes_loaded() == rhs.bytes_loaded() &&
+         lhs.documents_loaded() == rhs.documents_loaded() &&
+         lhs.total_bytes() == rhs.total_bytes() &&
+         lhs.total_documents() == rhs.total_documents();
+}
+
+inline bool operator!=(const LoadBundleTaskProgress lhs,
+                       const LoadBundleTaskProgress& rhs) {
+  return !(lhs == rhs);
+}
+
+/** A handle used to lookup and remove observer from the task. */
+using LoadBundleHandle = std::string;
+
+/** Observer type that is called by the task when there is an update. */
+using ProgressObserver = std::function<void(LoadBundleTaskProgress)>;
+
+/** Holds the `LoadBundleHandle` to `ProgressObserver` mapping. */
+using HandleObservers =
+    std::vector<std::pair<LoadBundleHandle, ProgressObserver>>;
+
+/**
+ * Represents the task of loading a Firestore bundle. It provides progress of
+ * bundle loading, as well as task completion and error events.
+ */
+class LoadBundleTask {
+ public:
+  explicit LoadBundleTask(std::shared_ptr<util::Executor> user_executor)
+      : user_executor_(std::move(user_executor)) {
+  }
+
+  /**
+   * Instructs the task to notify the specified observer when there is a
+   * progress update with the given `LoadBundleTaskState`.
+   *
+   * @return A handle that can be used to remove the callback from this task.
+   */
+  LoadBundleHandle ObserveState(LoadBundleTaskState state,
```

Comment:
They would only need to register for `kInProgress`.

As I am writing, it becomes clear this can be confusing. This is what the approved API is, so let's discuss with the team.

Reply:
Updated according to team discussion.